ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/state/PendingBlocksManager.java

```diff
@@ -27,35 +30,54 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
-public class PendingBlocks {
+import org.apache.tuweni.bytes.Bytes;
 
-  private final Map<Hash, Block> pendingBlocks = new ConcurrentHashMap<>();
+public class PendingBlocksManager {
+
+  private static final int REORG_CACHE_SIZE_FACTOR = 2;
+
+  private final PendingBlockCache pendingBlocks;
   private final Map<Hash, Set<Hash>> pendingBlocksByParentHash = new ConcurrentHashMap<>();
 
+  public PendingBlocksManager(final SynchronizerConfiguration synchronizerConfiguration) {
+    pendingBlocks =
+        new PendingBlockCache(
+            (Math.abs(synchronizerConfiguration.getBlockPropagationRange().lowerEndpoint())
+                    + Math.abs(
+                        synchronizerConfiguration.getBlockPropagationRange().upperEndpoint()))
+                * REORG_CACHE_SIZE_FACTOR);
```

Comment:
I'm not sure we need a REORG_CACHE_SIZE_MUTIPLICATOR. Couldn't we leave it the size of the inverval (40 in this case) and use a bounded priority queue? The highest priority blocks would be those that are lowest in block height and then higher priority if they were sent more recently.

Reply:
I don't think a queue is the best choice in this case because in my opinion we need to stay on a Map to be able to search for a block by its hash and not just browse the queue element by element. But a priority mechanism could be a good idea.  I can add this mechanism to my cache by creating a custom comparator and instead of refusing to add a block, I will replace the one that has the highest number and which is the older. What do you think ?

Reply:
Would a sorted map work? The comparator could be the block height and time we received it.

Reply:
I checked the SortedMap and knowing that the order must depend on the nodeId, the block Number and the timestamp. It would seem easier to get the lowest priority block using code like this when trying to add a new block (I need to verify that it works). 

```
private Optional<ImmutablePendingBlock> getLessPriorityBlock(
      final ImmutablePendingBlock pendingBlock) {
    

    // get all blocks from the nodeId
    final Stream<ImmutablePendingBlock> nodeIdBlocks =
        values().stream().filter(value -> value.nodeId() == pendingBlock.nodeId());

    // get the oldest block
    final Optional<ImmutablePendingBlock> oldestBlock =
        nodeIdBlocks.filter(value -> value.nodeId() == pendingBlock.nodeId())
            .max(Comparator.comparingLong(o -> o.block().getHeader().getNumber())).stream()
            .findFirst();
    
    // filter by number and timestamp
    return oldestBlock.flatMap(
        immutablePendingBlock ->
            nodeIdBlocks
                .filter(
                    block ->
                        block.block().getHeader().getNumber()
                            == immutablePendingBlock.block().getHeader().getNumber())
                .min(Comparator.comparingLong(o -> o.block().getHeader().getTimestamp())));
  }
```

In addition, the sorting on a SortedMap seems to be done only on the key during insertion. You have to browse the whole map if you want to sort by value.

Reply:
What do you think about using a queue anyways since it's going to be max of 40 blocks. O(n) with n=40 shouldn't be too bad when searching for hashes and it would make the code less complicated _I think_. Not sure.

Reply:
not necessarily 40 because if we are looking for a block we will not only search for a single peer but for all the peers so 40 * nbPeers. I will continue to search and see if another solution is possible. The initial idea of ​​having only a simple implementation that blocks the new blocks was precisely to not impact performance by sorting everytime. Because the case where we have more than 40 blocks will be relatively rare but the sort will be done each time.

Reply:
another solution would be to add a priority queue for each peer next to the global map. and when adding check in this queue. maybe that's the solution. it will just be necessary to work in the synchronization of these two component. But it doesn't seem complicated because there is already "pendingBlocksByParentHash" which is synchronized.

Reply:
I think I may have a way to simplify the code like this

````
private Optional<ImmutablePendingBlock> getLessPriorityBlock(
      final ImmutablePendingBlock pendingBlock) {
    Comparator<ImmutablePendingBlock> comparator = Comparator.comparing(s -> s.block().getHeader().getNumber());
    comparator = comparator.thenComparing(s->  s.block().getHeader().getTimestamp());
    return values().stream().filter(value -> value.nodeId() == pendingBlock.nodeId()).min(comparator);
  }
````